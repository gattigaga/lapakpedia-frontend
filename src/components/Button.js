import React from "react";
import PropTypes from "prop-types";
import styled from "styled-components";
import { Link } from "react-router-dom";

const StyledButton = styled.button`
  background: ${props => (props.isOutlined ? "white" : "black")};
  color: ${props => (props.isOutlined ? "black" : "white")};
  border: 1px solid black;
  outline: none;
  padding: 8px 16px;
  font-family: Roboto;
  cursor: pointer;
  transition: all 0.15s ease-in;
  letter-spacing: 0.1em;
  width: ${props => (props.isFullWidth ? "100%" : "auto")};

  &:hover {
    background: ${props => (props.isOutlined ? "black" : "white")};
    color: ${props => (props.isOutlined ? "white" : "black")};
  }
`;

const Button = ({ caption, type, onClick, href, isOutlined, isFullWidth }) => {
  const button = (
    <StyledButton
      type={type}
      onClick={onClick}
      isOutlined={isOutlined}
      isFullWidth={isFullWidth}
    >
      {caption}
    </StyledButton>
  );

  return href ? <Link to={href}>{button}</Link> : button;
};

Button.propTypes = {
  caption: PropTypes.string,
  type: PropTypes.string,
  onClick: PropTypes.func,
  href: PropTypes.string,
  isOutlined: PropTypes.bool,
  isFullWidth: PropTypes.bool
};

Button.defaultProps = {
  caption: "Button",
  type: "button"
};

export default Button;
