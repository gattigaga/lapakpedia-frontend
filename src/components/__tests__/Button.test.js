import React from "react";
import { shallow } from "enzyme";
import toJSON from "enzyme-to-json";

import Button from "components/Button";

describe("Button", () => {
  const setup = propOverrides => {
    const props = {
      ...propOverrides
    };

    const wrapper = shallow(<Button {...props} />);

    return {
      props,
      wrapper
    };
  };

  it("should renders default", () => {
    const { wrapper } = setup();

    expect(toJSON(wrapper)).toMatchSnapshot();
  });

  it("should renders with link", () => {
    const { wrapper } = setup({ href: "https://github.com" });

    expect(wrapper.find("Link").props().to).toEqual("https://github.com");
  });

  it("should renders with custom type", () => {
    const { wrapper } = setup({ type: "submit" });

    expect(wrapper.props().type).toEqual("submit");
  });

  it("should renders with custom caption", () => {
    const { wrapper } = setup({ caption: "Login" });

    expect(wrapper.props().children).toEqual("Login");
  });

  it("should renders with outline", () => {
    const { wrapper } = setup({ isOutlined: true });

    expect(wrapper.props().isOutlined).toEqual(true);
  });

  it("should renders in full width", () => {
    const { wrapper } = setup({ isFullWidth: true });

    expect(wrapper.props().isFullWidth).toEqual(true);
  });

  it("should call 'onClick' callback while clicked", () => {
    const onClick = jest.fn();
    const { wrapper } = setup({ onClick });

    wrapper.simulate("click");

    expect(onClick).toBeCalled();
  });
});
