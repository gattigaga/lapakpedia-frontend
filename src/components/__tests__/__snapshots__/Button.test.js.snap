// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button should renders default 1`] = `
<styled.button
  type="button"
>
  Button
</styled.button>
`;
